### ISSUE TITLE
Falcon code fails "strict" compilation

### ISSUE THREAD

**Issue by mouse07410** [PQClean/PQClean#355]
Lines
https://github.com/PQClean/PQClean/blob/3c8be3cb1f8ca0953f2df44ea665f118efb468d6/crypto_sign/falcon-1024/avx2/rng.c#L94
and
https://github.com/PQClean/PQClean/blob/3c8be3cb1f8ca0953f2df44ea665f118efb468d6/crypto_sign/falcon-1024/avx2/rng.c#L99
cause the GCC compiler to emit warning "`dereferencing type-punned pointer will break strict-aliasing rules`".

As we compile with `-Werror`, it breaks our build. This applies to both `avx2` and `clean` versions.

One possible fix is replacing those with something like `memcpy(&cc, (p->state.d + 48), 8);`

**Comment by jschanck** [comment]
To avoid the strict-aliasing warning we should access the memory with pointers that are declared as part of the unions. So we should change

https://github.com/PQClean/PQClean/blob/3c8be3cb1f8ca0953f2df44ea665f118efb468d6/crypto_sign/falcon-1024/avx2/inner.h#L462-L473

to

```c
typedef struct {
    union {
        uint8_t d[512]; /* MUST be 512, exactly */
        uint32_t d32[512/4];
        uint64_t d64[512/8];
    } buf;
    size_t ptr;
    union {
        uint8_t d[256];
        uint32_t d32[256/4];
        uint64_t d64[256/8];
    } state;
    int type;
} prng;
```

The alignment trick that is causing the warning is common throughout the PQClean codebase. So I think we should start by developing a test that catches all instances of this.

@mouse07410 What version of GCC are you using? And does it issue aliasing warnings in any schemes other than falcon?

**Comment by mouse07410** [comment]
> What version of GCC are you using? 

```
COLLECT_GCC=gcc
COLLECT_LTO_WRAPPER=/usr/libexec/gcc/x86_64-redhat-linux/4.8.5/lto-wrapper
Target: x86_64-redhat-linux
Configured with: ../configure --prefix=/usr --mandir=/usr/share/man --infodir=/usr/share/info --with-bugurl=http://bugzilla.redhat.com/bugzilla --enable-bootstrap --enable-shared --enable-threads=posix --enable-checking=release --with-system-zlib --enable-__cxa_atexit --disable-libunwind-exceptions --enable-gnu-unique-object --enable-linker-build-id --with-linker-hash-style=gnu --enable-languages=c,c++,objc,obj-c++,java,fortran,ada,go,lto --enable-plugin --enable-initfini-array --disable-libgcj --with-isl=/builddir/build/BUILD/gcc-4.8.5-20150702/obj-x86_64-redhat-linux/isl-install --with-cloog=/builddir/build/BUILD/gcc-4.8.5-20150702/obj-x86_64-redhat-linux/cloog-install --enable-gnu-indirect-function --with-tune=generic --with-arch_32=x86-64 --build=x86_64-redhat-linux
Thread model: posix
gcc version 4.8.5 20150623 (Red Hat 4.8.5-36) (GCC)
```

> And does it issue aliasing warnings in any schemes other than falcon?

Alas, I'm using PQClean in the "recommended" way - aka I take the scheme(s) that my project needs, and embed the source directories into it. I'll try Dilithium and NTRU with the project compiler.

So, I haven't tried other schemes - except in `pytest -v -n=auto` (where everything seems fine, with Clang-10 and GCC-10.2).

**Comment by mouse07410** [comment]
I still need to test Dilithium. NTRU is fine everywhere.

HQC showed a problem with GCC-10.2:
```
_____________________ test_symbol_namespaces[avx2 implementation of hqc-rmrs-256] ______________________
[gw3] darwin -- Python 3.8.6 /opt/local/Library/Frameworks/Python.framework/Versions/3.8/bin/python3.8

implementation = <Implementation(hqc-rmrs-256, avx2)>
impl_path = '/Users/ur20980/src/PQClean/testcases/test_symbol_ns__bZsmaOK7/crypto_bla/scheme/impl'
test_dir = '/Users/ur20980/src/PQClean/testcases/test_symbol_ns__bZsmaOK7'
init = <function isolate_test_files.<locals>.initializer at 0x10ab82940>
destr = <function isolate_test_files.<locals>.destructor at 0x10ab829d0>

    @pytest.mark.parametrize(
        'implementation,test_dir,impl_path,init,destr',
        [(impl,
          *helpers.isolate_test_files(impl.path(), 'test_symbol_ns_'))
         for impl in pqclean.Scheme.all_supported_implementations()],
        ids=[str(impl) for impl in pqclean.Scheme.all_supported_implementations()],
    )
    @helpers.filtered_test
    def test_symbol_namespaces(implementation, impl_path, test_dir, init, destr):
        if sys.platform not in ['linux', 'darwin']:
            raise unittest.SkipTest("Unsupported platform")
        init()
        helpers.make(working_dir=impl_path)
        out = helpers.run_subprocess(
            ['nm', '-g', implementation.libname()],
            impl_path,
        )
    
        lines = out.strip().split("\n")
        symbols = []
        for line in lines:
            if line.startswith("00"):
                symbols.append(line)
    
        namespace = implementation.namespace_prefix()
        non_namespaced = []
        for symbolstr in symbols:
            *_, symtype, symbol = symbolstr.split()
            if symtype in 'TDRS':
                if not (symbol.startswith(namespace) or
                        symbol.startswith('_' + namespace) or
                        # KeccakP-1600 for AVX2
                        symbol.startswith('KeccakF1600times4') or
                        symbol.startswith('_KeccakF1600times4') or  # MacOS
                        symbol.startswith('KeccakP1600times4') or
                        symbol.startswith('_KeccakP1600times4') or  # MacOS
                        # weird things on i386
                        symbol.startswith('__x86.get_pc_thunk.')):
                    non_namespaced.append(symbol)
    
        if non_namespaced:
            print("Missing namespace literal {}".format(namespace))
            for symbol in non_namespaced:
                print("\ttype: {}, symbol: {}".format(symtype, symbol))
    
>       assert not non_namespaced, "Literals with missing namespaces"
E       AssertionError: Literals with missing namespaces
E       assert not ['_bit64', '_bloc64']

test_symbol_namespace.py:60: AssertionError
----------------------------------------- Captured stdout call -----------------------------------------
/Users/ur20980/src/PQClean/testcases/test_symbol_ns__bZsmaOK7/crypto_bla/scheme/impl > make
gcc -O3 -mavx2 -mbmi -mpclmul -Wall -Wextra -Wshadow -Wpedantic -Wvla -Werror -Wredundant-decls -Wmissing-prototypes -std=c99 -I../../../common  -c -o code.o code.c
gcc -O3 -mavx2 -mbmi -mpclmul -Wall -Wextra -Wshadow -Wpedantic -Wvla -Werror -Wredundant-decls -Wmissing-prototypes -std=c99 -I../../../common  -c -o fft.o fft.c
gcc -O3 -mavx2 -mbmi -mpclmul -Wall -Wextra -Wshadow -Wpedantic -Wvla -Werror -Wredundant-decls -Wmissing-prototypes -std=c99 -I../../../common  -c -o gf2x.o gf2x.c
gcc -O3 -mavx2 -mbmi -mpclmul -Wall -Wextra -Wshadow -Wpedantic -Wvla -Werror -Wredundant-decls -Wmissing-prototypes -std=c99 -I../../../common  -c -o gf.o gf.c
gcc -O3 -mavx2 -mbmi -mpclmul -Wall -Wextra -Wshadow -Wpedantic -Wvla -Werror -Wredundant-decls -Wmissing-prototypes -std=c99 -I../../../common  -c -o hqc.o hqc.c
gcc -O3 -mavx2 -mbmi -mpclmul -Wall -Wextra -Wshadow -Wpedantic -Wvla -Werror -Wredundant-decls -Wmissing-prototypes -std=c99 -I../../../common  -c -o kem.o kem.c
gcc -O3 -mavx2 -mbmi -mpclmul -Wall -Wextra -Wshadow -Wpedantic -Wvla -Werror -Wredundant-decls -Wmissing-prototypes -std=c99 -I../../../common  -c -o parsing.o parsing.c
gcc -O3 -mavx2 -mbmi -mpclmul -Wall -Wextra -Wshadow -Wpedantic -Wvla -Werror -Wredundant-decls -Wmissing-prototypes -std=c99 -I../../../common  -c -o reed_muller.o reed_muller.c
gcc -O3 -mavx2 -mbmi -mpclmul -Wall -Wextra -Wshadow -Wpedantic -Wvla -Werror -Wredundant-decls -Wmissing-prototypes -std=c99 -I../../../common  -c -o reed_solomon.o reed_solomon.c
gcc -O3 -mavx2 -mbmi -mpclmul -Wall -Wextra -Wshadow -Wpedantic -Wvla -Werror -Wredundant-decls -Wmissing-prototypes -std=c99 -I../../../common  -c -o vector.o vector.c
ar -r libhqc-rmrs-256_avx2.a code.o fft.o gf2x.o gf.o hqc.o kem.o parsing.o reed_muller.o reed_solomon.o vector.o 
ar: creating archive libhqc-rmrs-256_avx2.a

/Users/ur20980/src/PQClean/testcases/test_symbol_ns__bZsmaOK7/crypto_bla/scheme/impl > nm -g libhqc-rmrs-256_avx2.a

libhqc-rmrs-256_avx2.a(code.o):
00000000000000a0 T _PQCLEAN_HQCRMRS256_AVX2_code_decode
0000000000000000 T _PQCLEAN_HQCRMRS256_AVX2_code_encode
                 U _PQCLEAN_HQCRMRS256_AVX2_reed_muller_decode
                 U _PQCLEAN_HQCRMRS256_AVX2_reed_muller_encode
                 U _PQCLEAN_HQCRMRS256_AVX2_reed_solomon_decode
                 U _PQCLEAN_HQCRMRS256_AVX2_reed_solomon_encode

libhqc-rmrs-256_avx2.a(fft.o):
0000000000001de0 T _PQCLEAN_HQCRMRS256_AVX2_fft
00000000000023b0 T _PQCLEAN_HQCRMRS256_AVX2_fft_retrieve_error_poly
                 U _PQCLEAN_HQCRMRS256_AVX2_gf_inverse
                 U _PQCLEAN_HQCRMRS256_AVX2_gf_mul
                 U _PQCLEAN_HQCRMRS256_AVX2_gf_square
                 U ___memcpy_chk
                 U _memcpy

libhqc-rmrs-256_avx2.a(gf2x.o):
00000000000021b0 T _PQCLEAN_HQCRMRS256_AVX2_vect_mul
0000000000002ae0 S _bit64
0000000000002fa0 S _bloc64
                 U _memcpy
                 U _memset

libhqc-rmrs-256_avx2.a(gf.o):
00000000000002e0 T _PQCLEAN_HQCRMRS256_AVX2_gf_inverse
00000000000008e0 T _PQCLEAN_HQCRMRS256_AVX2_gf_mod
0000000000000000 T _PQCLEAN_HQCRMRS256_AVX2_gf_mul
0000000000000070 T _PQCLEAN_HQCRMRS256_AVX2_gf_mul_vect
0000000000000230 T _PQCLEAN_HQCRMRS256_AVX2_gf_square

libhqc-rmrs-256_avx2.a(hqc.o):
                 U _PQCLEAN_HQCRMRS256_AVX2_code_decode
                 U _PQCLEAN_HQCRMRS256_AVX2_code_encode
0000000000000460 T _PQCLEAN_HQCRMRS256_AVX2_hqc_pke_decrypt
0000000000000220 T _PQCLEAN_HQCRMRS256_AVX2_hqc_pke_encrypt
0000000000000000 T _PQCLEAN_HQCRMRS256_AVX2_hqc_pke_keygen
                 U _PQCLEAN_HQCRMRS256_AVX2_hqc_public_key_from_string
                 U _PQCLEAN_HQCRMRS256_AVX2_hqc_public_key_to_string
                 U _PQCLEAN_HQCRMRS256_AVX2_hqc_secret_key_from_string
                 U _PQCLEAN_HQCRMRS256_AVX2_hqc_secret_key_to_string
                 U _PQCLEAN_HQCRMRS256_AVX2_load8_arr
                 U _PQCLEAN_HQCRMRS256_AVX2_store8_arr
                 U _PQCLEAN_HQCRMRS256_AVX2_vect_add
                 U _PQCLEAN_HQCRMRS256_AVX2_vect_mul
                 U _PQCLEAN_HQCRMRS256_AVX2_vect_resize
                 U _PQCLEAN_HQCRMRS256_AVX2_vect_set_random
                 U _PQCLEAN_HQCRMRS256_AVX2_vect_set_random_fixed_weight
                 U _memcpy
                 U _memset
                 U _randombytes
                 U _seedexpander_init

libhqc-rmrs-256_avx2.a(kem.o):
0000000000000220 T _PQCLEAN_HQCRMRS256_AVX2_crypto_kem_dec
0000000000000010 T _PQCLEAN_HQCRMRS256_AVX2_crypto_kem_enc
0000000000000000 T _PQCLEAN_HQCRMRS256_AVX2_crypto_kem_keypair
                 U _PQCLEAN_HQCRMRS256_AVX2_hqc_ciphertext_from_string
                 U _PQCLEAN_HQCRMRS256_AVX2_hqc_ciphertext_to_string
                 U _PQCLEAN_HQCRMRS256_AVX2_hqc_pke_decrypt
                 U _PQCLEAN_HQCRMRS256_AVX2_hqc_pke_encrypt
                 U _PQCLEAN_HQCRMRS256_AVX2_hqc_pke_keygen
                 U _PQCLEAN_HQCRMRS256_AVX2_store8_arr
                 U _PQCLEAN_HQCRMRS256_AVX2_vect_compare
                 U _memcpy
                 U _memset
                 U _randombytes
                 U _sha3_512
                 U _sha512

libhqc-rmrs-256_avx2.a(parsing.o):
0000000000000500 T _PQCLEAN_HQCRMRS256_AVX2_hqc_ciphertext_from_string
0000000000000400 T _PQCLEAN_HQCRMRS256_AVX2_hqc_ciphertext_to_string
0000000000000380 T _PQCLEAN_HQCRMRS256_AVX2_hqc_public_key_from_string
0000000000000300 T _PQCLEAN_HQCRMRS256_AVX2_hqc_public_key_to_string
0000000000000270 T _PQCLEAN_HQCRMRS256_AVX2_hqc_secret_key_from_string
0000000000000230 T _PQCLEAN_HQCRMRS256_AVX2_hqc_secret_key_to_string
00000000000000d0 T _PQCLEAN_HQCRMRS256_AVX2_load8
0000000000000130 T _PQCLEAN_HQCRMRS256_AVX2_load8_arr
00000000000000c0 T _PQCLEAN_HQCRMRS256_AVX2_store8
00000000000001c0 T _PQCLEAN_HQCRMRS256_AVX2_store8_arr
                 U _PQCLEAN_HQCRMRS256_AVX2_vect_set_random
                 U _PQCLEAN_HQCRMRS256_AVX2_vect_set_random_fixed_weight
                 U _memcpy
                 U _seedexpander_init

libhqc-rmrs-256_avx2.a(reed_muller.o):
0000000000000150 T _PQCLEAN_HQCRMRS256_AVX2_reed_muller_decode
0000000000000000 T _PQCLEAN_HQCRMRS256_AVX2_reed_muller_encode

libhqc-rmrs-256_avx2.a(reed_solomon.o):
                 U _PQCLEAN_HQCRMRS256_AVX2_fft
                 U _PQCLEAN_HQCRMRS256_AVX2_fft_retrieve_error_poly
                 U _PQCLEAN_HQCRMRS256_AVX2_gf_inverse
                 U _PQCLEAN_HQCRMRS256_AVX2_gf_mul
                 U _PQCLEAN_HQCRMRS256_AVX2_gf_mul_vect
0000000000000140 T _PQCLEAN_HQCRMRS256_AVX2_reed_solomon_decode
0000000000000000 T _PQCLEAN_HQCRMRS256_AVX2_reed_solomon_encode

libhqc-rmrs-256_avx2.a(vector.o):
                 U _PQCLEAN_HQCRMRS256_AVX2_load8_arr
0000000000000740 T _PQCLEAN_HQCRMRS256_AVX2_vect_add
0000000000000820 T _PQCLEAN_HQCRMRS256_AVX2_vect_compare
0000000000000860 T _PQCLEAN_HQCRMRS256_AVX2_vect_resize
00000000000006d0 T _PQCLEAN_HQCRMRS256_AVX2_vect_set_random
0000000000000000 T _PQCLEAN_HQCRMRS256_AVX2_vect_set_random_fixed_weight
                 U _memcpy
                 U _memset
                 U _seedexpander

Missing namespace literal PQCLEAN_HQCRMRS256_AVX2_
	type: T, symbol: _bit64
	type: T, symbol: _bloc64
======================================= short test summary info ========================================
FAILED test_symbol_namespace.py::test_symbol_namespaces[avx2 implementation of hqc-rmrs-256] - Assert...
======================= 1 failed, 4271 passed, 543 skipped in 5386.42s (1:29:46) =======================
```

**Comment by mouse07410** [comment]
@jschanck based on https://github.com/PQClean/PQClean/issues/355#issuecomment-729757527 (i.e., assuming the struct is extended as you suggested), how would you change this code (what should it look like):
```c
   cc = *(uint64_t *)(p->state.d + 48);
```

**Comment by jschanck** [comment]
```c
cc = p->state.d64[6]
```

**Comment by mouse07410** [comment]
@jschanck thanks - now the only problematic one seems to be 
https://github.com/PQClean/PQClean/blob/d32b1cd7c21c1411c85b2d4f5269a0aa412959fb/crypto_sign/falcon-1024/clean/rng.c#L71

When I changed it to 
```c
    p->state.d32[ (i << 2) ] = w;
```
the code started to fail test-vectors...

**Comment by jschanck** [comment]
If `x` has type `T *` then arithmetic with `x` has a stride of `sizeof T` bytes. So you just want

```c
    p->state.d32[i] = w;
```
or equivalently
```c
    *(p->state.d32 + i) = w;
```

**Comment by mouse07410** [comment]
> So you just want . . .

*\<Blushing\>* 
I should be ashamed of myself for asking such questions. Probably been working too hard before the Holidays. ;-)

Thank you - everything works now as it should.

Looking forward to having these changes incorporated in the main repo.

**Comment by thomwiggers** [comment]
This was originally fixed by #280, but an update reverted these changes. I will have another look.

**Comment by thomwiggers** [comment]
Should be fixed by #476
